### ISSUE TITLE
Serialized hyperparams break containers that aren't launched by python sdk

### ISSUE THREAD

**Issue by bweigel** [aws/sagemaker-containers#179]
Cheers,

I have already offloaded this to AWS support, but find it might help other people too, so here I go...

#### Problem

> Serialized hyperparams break containers that aren't launched by python sdk

We are currently using step-functions to automate training of some models using sagemaker. 
We use `arn:aws:states:::sagemaker:createTrainingJob.sync` as outlined [here](https://docs.aws.amazon.com/step-functions/latest/dg/connectors-sagemaker.html).

This in turn uses the sagemaker API as described [here](https://docs.aws.amazon.com/sagemaker/latest/dg/API_CreateTrainingJob.html#API_CreateTrainingJob_RequestSyntax), where it is specified to use `Hyperparameters` as follows:

```json
"HyperParameters": { 
      "string" : "string" 
   }
```

(Type: String to string map).

It is reasonable to expect the following syntax to be valid for `HyperParameters`:

```json
"HyperParameters": {
    "epochs": "1",
    "batch_size": "128",
    "conv_block_length": "2",
    "cycle_length": "10",
    "depth": "5",
    "dropout": "0.5",
    "job_name": "my_job",
    "max_lr": "0.1",
    "min_lr": "0.0001",
    "sagemaker_container_log_level": "20",
    "sagemaker_enable_cloudwatch_metrics": "false",
    "sagemaker_job_name": "my_job",
    "sagemaker_program": "train_n_folds.py",
    ...
}
```

However, this will fail the training job with the following error:

```
Traceback (most recent call last):
File "/usr/local/lib/python3.5/dist-packages/container_support/training.py", line 31, in start
env = TrainingEnvironment()
File "/usr/local/lib/python3.5/dist-packages/container_support/environment.py", line 182, in __init__
os.path.join(self.input_config_dir, TrainingEnvironment.HYPERPARAMETERS_FILE))
File "/usr/local/lib/python3.5/dist-packages/container_support/environment.py", line 224, in _load_hyperparameters
return self._deserialize_hyperparameters(serialized)
File "/usr/local/lib/python3.5/dist-packages/container_support/environment.py", line 238, in _deserialize_hyperparameters
hyperparameter_dict[k] = json.loads(v)
File "/usr/lib/python3.5/json/__init__.py", line 319, in loads
return _default_decoder.decode(s)
File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/lib/python3.5/json/decoder.py", line 357, in raw_decode
raise JSONDecodeError("Expecting value", s, err.value) from None
```

After some digging what caused this issue I became frustrated, because I couldn't find the code from the above stack-trace anywhere on GitHub. (:thinking: Am I just looking in the wrong place, or is this intended? If intended, will it ever be published for input from the community? :thinking:)

To get access to the code I downloaded the docker image from ECR and copied it from there and lo and behold I found something:

```python
class TrainingEnvironment(ContainerEnvironment):

    ...

    # TODO expecting serialized hyperparams might break containers that aren't launched by python sdk
    @staticmethod
    def _deserialize_hyperparameters(hp):
        hyperparameter_dict = {}

        for (k, v) in hp.items():
            # Tuning jobs inject a hyperparameter that does not conform to the JSON format
            if k == '_tuning_objective_metric':
                if v.startswith('"') and v.endswith('"'):
                    v = v.strip('"')
                hyperparameter_dict[k] = v
            else:
                hyperparameter_dict[k] = json.loads(v)

        return hyperparameter_dict
```

So somebody had the right idea :+1:, however it wasn't followed trough with :-1: .

What seems to break things is the fact that `json.loads(v)` is called on each element of the dict. However, if that element is a string (as expected by a string to string map) it will raise a `JSONDecodeError`. The sagemaker-sdk will do a `json.dumps("value")` for every input, thus yielding something like this `'"mystring"'`. You can do a `json.dumps('"mystring"')`, but a `json.dumps("mystring")` will fail.

#### Solution

There are some possibilities:
1. open source the code for people to help (if this hasn't happened)
2. fix the TODO
3. ... use an ugly workaround (see below)

**Workaround**

Our current workaround is quoting the "true" strings explicitly:

```json
"HyperParameters": {
    "epochs": "1",
    "batch_size": "128",
    "conv_block_length": "2",
    "cycle_length": "10",
    "depth": "5",
    "dropout": "0.5",
    "job_name": "\"my_job\"",
    "max_lr": "0.1",
    "min_lr": "0.0001",
    "sagemaker_container_log_level": "20",
    "sagemaker_enable_cloudwatch_metrics": "false",
    "sagemaker_job_name": "\"my_job\"",
    "sagemaker_program": "\"train_n_folds.py\"",
    ...
}
```

**Comment by mvsusp** [comment]
Hello @bweigel,

Thanks for the detailed issue report.  The newest versions of SageMaker Containers work exactly like you suggested: https://github.com/aws/sagemaker-containers/blob/master/src/sagemaker_containers/_env.py#L197

Which container version are you using? Is it possible to update it to a newest version so you are able to use the code changes?

Thanks for using SageMaker!

**Comment by bweigel** [comment]
Thanks for the reply @mvsusp! We are currently using the image  `sagemaker-mxnet:1.2-gpu-py3`. We will check out the new version 1.3 and see if it works for us, but it looks like a major overhaul in the support code. 

I will get back, once we have evaluated this.

**Comment by laurenyu** [comment]
@bweigel is this still an issue for you? Were you able to use the MXNet 1.3 images?

the code you referenced is open source already - it's in the r1.0 branch of this repository. As Marcio mentioned, we've been migrating our Docker images to use the updated version of this repository.

**Comment by bweigel** [comment]
Sorry to have kept you waiting. This issue is fixed in MXNet 1.3 images. Thank you :relaxed:

**Comment by laurenyu** [comment]
glad the MXNet 1.3 images worked out for you!
